spmd/tensor/parallel/api.py

```diff
@@ -1,122 +1,194 @@
                 )
-                module.register_parameter("weight", sharded_weight)
-                if module.bias is not None:
-                    replicated_bias = nn.Parameter(
-                        distribute_tensor(module.bias, device_mesh, replicate)
-                    )
-                    module.register_parameter("bias", replicated_bias)
-
-    if isinstance(module, TensorParallelMultiheadAttention):  # shard TPMA
-        for n, m in module.named_children():
-            _shard_self_attn_params(n, m)
+            )
+            return module
     else:
-        for n, m in module.named_children():  # replace with TPMA
-            if isinstance(m, nn.MultiheadAttention):
-                tp_multi_head_attention = TensorParallelMultiheadAttention(
-                    m.embed_dim,
-                    m.num_heads,
-                    device=torch.device(device_mesh.device_type),
-                    tp_size=device_mesh.size(0),  # group size on dim 0
-                    add_bias_kv=m.bias_k is not None,
-                )
-                tp_multi_head_attention.copy(m)
-                module.register_module(n, tp_multi_head_attention)
+        raise RuntimeError(  # pyre-ignore[7]
+            f"Expect Union[ParallelStyle, Dict[str, ParallelStyle]] for parallelize_plan, {type(parallelize_plan)} found!"
+        )
 
 
-def _has_even_num_linears(module: nn.Module) -> bool:
+def _is_mha_for_pairwise_parallel(module: nn.Module) -> bool:
     """
-    We traverse through all the children of the given module and count the
-    number of Linear module. If the number is even, we return True.
+    Check whether the mha module is the one can be handled for Pairwise parallel.
+
+    Args:
+        module (nn.Module):
+            :class:``nn.Module`` object to be checked.
+
+    Return:
+        A boolean object which specifies whether the module is MHA supported by Pairwise parallel or not.
+    """
+    return isinstance(module, TensorParallelMultiheadAttention) or isinstance(
+        module, nn.MultiheadAttention
+    )
+
+
+def _is_mlp_for_pairwise_parallel(module: nn.Module) -> bool:
+    """
+    Traverse through all the immediate children of the given module and count the
+    number of Linear module. If the number is more than one, we return True.
 
     Args:
         module (nn.Module):
             :class:``nn.Module`` object to be traversed and counted.
 
     Return:
-        A boolean object which specifies whether the module contains
-        event-number of Linears in its children.
+        A boolean object which specifies whether the module is MLP or not.
 
     .. warning::
         The traversal is not recursive for now.
     """
     linear_submodules = list(
         filter(lambda x: isinstance(x, nn.Linear), module.children())
     )
-    return len(linear_submodules) > 0 and len(linear_submodules) % 2 == 0
+    return len(linear_submodules) > 0
```

Comment:
On a second thought, shouldn't we return when `len(linear_submodules) > 1` instead? It also matches with the error throwing below.